### DISCUSSION TITLE
New page insertion failure

### DISCUSSION THREAD

**Discussion by qwertynik** [pymupdf/PyMuPDF#1732] [Looking for help] [answered]
With the below snippet the attempt is to insert two pages - one after another.

```
        pdf = fitz.open()
        pageWidth = 600
        pageHeight = 1000

        # create a new page from the PDF.
        firstPage = pdf.new_page(width=pageWidth, height=pageHeight)

        newPage = pdf.new_page(width=firstPage.rect.width - 100, height=firstPage.rect.height - 100, pno=pdf.page_count)

        print(firstPage)
```

But, the statement `print(firstPage)` leads to this error:
```
Error
Traceback (most recent call last):
  File "C:\Users\syste\PycharmProjects\test\test\pdf\test_Pdf.py", line 35, in test_consecutive_pdf_page_creation
    print(firstPage)
  File "C:\Users\syste\PycharmProjects\test\venv\lib\site-packages\fitz\fitz.py", line 6730, in __str__
    CheckParent(self)
  File "C:\Users\syste\PycharmProjects\test\venv\lib\site-packages\fitz\fitz.py", line 2802, in CheckParent
    raise ValueError("orphaned object: parent is None")
ValueError: orphaned object: parent is None
```
Different variations attempted for the `pno=pdf.page_count` param:
1. No param passed.
2. -1 passed.
3. pdf.page_count

All of them return the same error. Had re-read the documentation no clarity yet though. @JorjMcKie Any ideas on how to resolve this?

**Comment by JorjMcKie** [comment]
Page number is a **_positional_** parameter with a default "after last page".

Otherwise, I don't know yet why the previous page object is disabled even when the new page is created with a different name.

Circumvent this issue by
1. either loading an old page again if you must access any of its information, e.g. `doc[-1].rect.width`.
2. or append previous pages to some list and access them by index
3. or access any page dimension directly from the page object definition, like `doc.get_page_cropbox(pno)` which will not load the page (in your case the cropbox equals `page.rect`).

Methods 1. and 2. are highly inefficient and a waste of resources. A variant of method 2 would be to just store the information you may need later, or maybe a `dict`, so you cab do this: `page_rect[pno]`.

**Comment by qwertynik** [comment]
@JorjMcKie 
> Page number is a positional parameter with a default "after last page".
> 
> Otherwise, I don't know yet why the previous page object is disabled even when the new page is created with a different name.

Yes, this is perplexing. Most likely a bug in the library.

>2. or append previous pages to some list and access them by index

Like done in the code sample pasted, Isn't this similar to storing the page's reference to a variable?

>3. or access any page dimension directly from the page object definition, like doc.get_page_cropbox(pno) which will not load the page (in your case the cropbox equals page.rect).

Wondering how this will work, because, this will most likely access page's reference in the doc object which would same reference as the variable.

Assigment statement is straight-forward, however doesn't work as intended. Is it due to not knowing the underlying guidelines and principles using which the library is built?

**Reply by JorjMcKie** [reply]
As I wrote: `doc.get_page_cropbox(pno)` does work and return the page rectangle of page number `pno`. It does not created the resp. (Python) page object, instead looks up the page's PDF object definition.

**Reply by qwertynik** [reply]
@JorjMcKie Yes, had noted the same. However, the goal here is to read the rect properties alone. But also work with the previous pages. Any ideas on how can this be accomplished?

**Reply by JorjMcKie** [reply]
you can always reload an existing page like so:
```python
doc=fitz.open()
page=doc.new_page(width=pw, height=ph)
for i in range(5):
    oldpage=doc[-1]
    page=doc.new_page(width=oldpage.rect.width-50, height=oldpage.rect.height-50)

    
for page in doc:
    print(tuple(page.rect))

    
(0.0, 0.0, 1000.0, 1000.0)
(0.0, 0.0, 950.0, 950.0)
(0.0, 0.0, 900.0, 900.0)
(0.0, 0.0, 850.0, 850.0)
(0.0, 0.0, 800.0, 800.0)
(0.0, 0.0, 750.0, 750.0)
```

**Reply by qwertynik** [reply]
Yes, this does work. Thanks!

**Comment by JorjMcKie** [comment]
> Wondering how this will work, because, this will most likely access page's reference in the doc object which would same reference as the variable.

No! `doc.get_page_cropbox(pno)` creates / accesses no Python object but accesses properties of the **_PDF_** object associated with the page's xref. A fitz `Document` contains no reference to all of its pages - that would be ridiculous.
Only the currently loaded ones are being referenced by it.

**Reply by qwertynik** [reply]
> A fitz Document contains no reference to all of its pages - that would be ridiculous.
> Only the currently loaded ones are being referenced by it.

Ok. So probably, post page creation 'older' pages are unloaded due to which a page isn't accessible. Not knowing this, was what led to the confusion of incorrectly expecting a Page object.

**Comment by JorjMcKie** [comment, marked as answer]
Found the bug. The line marked yellow is wrong and should be deleted. It deletes (disables) any previously loaded page:
![grafik](https://user-images.githubusercontent.com/8290722/170821015-9511b481-f428-4186-8dd6-c3479dc2ac0c.png)

**Reply by qwertynik** [reply]
Good that it has been found 👍
